### PR TITLE
Add a Power sensor for inverter diagnostics

### PR DESCRIPTION
**Pull request by hufman**
A draft of an energy sensor to expose the data from https://github.com/PeteRager/lennoxs30api/pull/23

image

### CODE REVIEW

**Comment by PeteRager** [comment]
Does this work at diagLevel 0 or do we need a way to set the diag level?

File: custom_components/lennoxs30/sensor.py [hunk 1]
```diff
@@ -47,6 +49,10 @@ async def async_setup_platform(
         _LOGGER.info(f"Create S30OutdoorTempSensor sensor system [{system.sysId}]")
         sensor = S30OutdoorTempSensor(hass, manager, system)
         sensor_list.append(sensor)
+        if system.diagInverterInputVoltage != None:
```
Comment (PeteRager) [inline custom_components/lennoxs30/sensor.py:52]: Let’s add a config option to enable.    Not all HVACs have inverters just the variable speed ones, so this sensor will only be relevant for those users.  Also, as a preference, I like an opt-in model prior to creating new sensors.

File: custom_components/lennoxs30/sensor.py [hunk 2]
```diff
@@ -242,3 +248,55 @@ def unit_of_measurement(self):
...
+        return False
+
+    @property
+    def name(self):
+        return self._myname
+
+    @property
+    def state(self):
+        try:
+            return float(self._system.diagInverterInputVoltage) * float(self._system.diagInverterInputCurrent)
```
Comment (PeteRager) [inline custom_components/lennoxs30/sensor.py:291]: Should we try to control the precision here.  For example, reporting 70.0023 watts then 70.0024 watts would not be ideal.

Rounding to the nearest Watt may be sufficient?  1/10 watt?

**Review by PeteRager — CHANGES_REQUESTED** [review]

File: custom_components/lennoxs30/sensor.py (hunk 2 above)
Comment (hufman) [reply custom_components/lennoxs30/sensor.py:291]: Very good point, integer-precision is probably plenty for this range (70-1500W from what I've seen so far)

File: custom_components/lennoxs30/sensor.py (hunk 1 above)
Comment (hufman) [reply custom_components/lennoxs30/sensor.py:52]: Can do!

**Comment by hufman** [comment]
> Does this work at diagLevel 0 or do we need a way to set the diag level?

It works at diag level 1 or 2, and the default diag level 0 does not emit the messages with the information.

**Review by PeteRager — APPROVED** [review]

**Comment by PeteRager** [comment]
Thanks for the changes and doc updates!

A question - is how is the diagnostic level set?  Is this done through the thermostat GUI (advanced settings) or do we need to provide an API command to set the diagLevel?  Then in HA do we set the diagLevel in the Config or expose a method?  We can make a new issue to track this.

semi-related - have you seen anymore S30s reboots?

**Comment by hufman** [comment]
Indeed, I planned the diagnostic setting to be a separate PR. I did accidentally discover a local menu to set it, not sure if everyone has it though:
1. Click the corner button and then Settings
2. Click "advanced settings" and then "view dealer control center"
3. Click "diagnostics", and then "Start Diagnostics"
4. It now shows me all of the diagnostics that I recognize from the diagnostic status messages

I think you can back out of this menu without needing to clock the Stop Diagnostics button.

I did see one reboot yesterday, while trying out the diagnostic levels at 0, 1, and 2 with my test script.

**Comment by PeteRager** [comment]
We have a choice on that PR to implement / automate the diagnostic setting or document how to do it manually.  Does the UI set it in Level 2?   

Sounds like we may have a repro for how to have the S30 reboot.   File a ticket with Lennox 😀

We should run a stability test for a week with diags turned on.  With diags and running local mine has been stable for 8 days.  

Does the diag setting get cleared on reboot?

**Comment by hufman** [comment]
Yes I believe the diagnostic setting gets reset back to off during a reboot. It might also reset after a timeout of a day or two, my graphs keep dropping out after a while.

**Comment by PeteRager** [comment]
> Yes I believe the diagnostic setting gets reset back to off during a reboot. It might also reset after a timeout of a day or two, my graphs keep dropping out after a while.

Perhaps what we should do is have a method in the API to set the preferred diagnostic level.  At the location we copy the diagLevel from status message into attribute, we could compare it against the preferred and spin off an async task to set it back.   The alternate implementation would be to do this in the integration by registering a callback whenever diagLevel changes and then spin off an async task to set it back.

**Comment by PeteRager** [comment]
I guess that is all dependent on the S30 properly sending us the diagLevel when it gets reset.  The should appear in debug logs.  It’s also mapped as an attribute in the Climate entity.  So we could make a sensor template and add it to the recorder to track what is happening

**Pull request merged**